src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -0,0 +1,144 @@
+#include "cluster_task_manager.h"
+
+#include "ray/util/logging.h"
+
+namespace ray {
+namespace raylet {
+
+ClusterTaskManager::ClusterTaskManager(
+    const ClientID &self_node_id,
+    std::shared_ptr<ClusterResourceScheduler> cluster_resource_scheduler,
+    std::function<bool(const Task &)> fulfills_dependencies_func,
+    std::shared_ptr<gcs::GcsClient> gcs_client)
+    : self_node_id_(self_node_id),
+      cluster_resource_scheduler_(cluster_resource_scheduler),
+      fulfills_dependencies_func_(fulfills_dependencies_func),
+      gcs_client_(gcs_client) {}
+
+bool ClusterTaskManager::SchedulePendingTasks() {
+  size_t queue_size = tasks_to_schedule_.size();
+  bool did_schedule = false;
+
+  // Check every task in task_to_schedule queue to see
+  // whether it can be scheduled. This avoids head-of-line
+  // blocking where a task which cannot be scheduled because
+  // there are not enough available resources blocks other
+  // tasks from being scheduled.
+  while (queue_size-- > 0) {
+    Work work = tasks_to_schedule_.front();
+    tasks_to_schedule_.pop_front();
+    Task task = work.second;
+    auto request_resources =
+        task.GetTaskSpecification().GetRequiredResources().GetResourceMap();
+    int64_t violations = 0;
+    std::string node_id_string = cluster_resource_scheduler_->GetBestSchedulableNode(
+        request_resources, &violations);
+    if (node_id_string.empty()) {
+      /// There is no node that has available resources to run the request.
+      tasks_to_schedule_.push_back(work);
+      continue;
+    } else {
+      if (node_id_string == self_node_id_.Binary()) {
+        did_schedule = did_schedule || WaitForTaskArgsRequests(work);
+      } else {
+        // Should spill over to a different node.
+        cluster_resource_scheduler_->AllocateRemoteTaskResources(node_id_string,
+                                                                 request_resources);
+
+        ClientID node_id = ClientID::FromBinary(node_id_string);
+        auto node_info_opt = gcs_client_->Nodes().Get(node_id);
+        RAY_CHECK(node_info_opt)
+            << "Spilling back to a node manager, but no GCS info found for node "
+            << node_id;
+        work.first(nullptr, node_id, node_info_opt->node_manager_address(),
+                   node_info_opt->node_manager_port());
+      }
+    }
+  }
+  return did_schedule;
+}
+
+bool ClusterTaskManager::WaitForTaskArgsRequests(Work work) {
+  Task task = work.second;
+  auto object_ids = task.GetTaskSpecification().GetDependencies();
+  bool can_dispatch = true;
+  if (object_ids.size() > 0) {
+    bool args_ready = fulfills_dependencies_func_(task);
+    if (args_ready) {
+      tasks_to_dispatch_.push_back(work);
+    } else {
+      can_dispatch = false;
+      TaskID task_id = task.GetTaskSpecification().TaskId();
+      waiting_tasks_.try_emplace(task_id, work);
```

Comment:
Is it every possible for the emplace to fail? Might want to make this an assertion.